### PR TITLE
Add --warning to suppress deprecation warnings (#3726)

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

This is a first-step implementation of #3728 . More design spec please see https://docs.google.com/document/d/1x5pasO5z5wZWLJBKHbgVbEsSX60jSv9s3uJn8e8bwQc/edit

In short, we don't want users to see deprecation warnings by default. This PR suppresses all warnings by default and adds `--warnings` to control this behavior:

- No option. By default don't display all warnings and render a summary at the end of the build: "There're {} warnings in your build, please run --warnings=all to see them."
- `--warnings=none` Don't display warnings and summary.
- `--warnings=all` Just like what Gradle does now.

This PR removes the refactoring part in #3751 to make review a bit easier.

### CODE REVIEW

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java [hunk 1]
```diff
@@ -44,6 +47,21 @@
      */
     void setConsoleOutput(ConsoleOutput consoleOutput);
 
+    /**
+     * Specifies if deprecation warnings should be written to the console.
+     * Defaults to {@link WarningType#Summary}
+     * @since 4.5
+     */
+    @Incubating
+    Set<WarningType> getWarningTypes();
```
Comment (wolfs) [inline subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I somewhat don't understand how `NONE` fits into a set of warnings. Would that mean that, if I use the invocation
```
./gradlew --warnings all --warnings none
```
then this set is `[NONE, ALL]` and all warnings would be displayed?
I also wouldn't be able to override a setting of `org.gradle.warnings=all` in my `gradle.properties` by passing `--warnings none` on the command line, right?

Or am I understanding something wrong here?

Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: Well, in my understanding, this should be "undefined behavior" currently. This usage is confusing for user, just like `./gradlew --info --debug --error`. Currently we only have `--all` and `--none`, they shouldn't show up at the same time because it's mutually exclusive. In the future, it could be something like `--warnings deprecations --warnings compiler`.

Here we use `Set` because this is a public API, we should not change it after publication.

Comment (wolfs) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I think the use case of setting `org.gradle.warnings=all` in `gradle.properties` and then running with `--warnings none` is a legitimate use case.

I think that `deprecation` and `compiler` are something completely different to `all`, `none` and `summary`. The latter three determine what to which relevant warnings, i.e. show them individually, don't show them, or show their number.
`deprecation` and `compiler` on the other hand are which types of warnings to take into account.

So I guess something like `--show-warnings all,summary` makes sense as would `--show-warnings summary --warnings compiler,deprecations`.

I guess `none` would evaluate to an empty set for how to show the warnings instead of a set with the element `NONE`. `all` would then probably be the set with `INDIVIDUAL,SUMMARY` (if that is what we want).

WDYT?

Comment (oehme) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I think we could do what the java compiler does. It's one option with 2 special values ("all" and "none"). The special values overwrite anything that was there before. The other values are simply additive.

See https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html#xlintwarnings

Comment (wolfs) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: @oehme That sounds good for `all` or `none` but it doesn't fit with `summary` I guess.
The Java compiler also uses categories which can be activated by `-Xlint:all`, `-Xlint:none` or individually by `-Xlint:<category>`, but not different kinds of reporting (e.g. our `summary`).

It uses `-Xmaxwarns number` to select the number of warnings to show. That is similar to our `summary` option I guess.

Comment (oehme) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I see what you mean, that split makes sense.

File: subprojects/core-api/src/main/java/org/gradle/StartParameter.java [hunk 2]
```diff
@@ -144,6 +145,16 @@ public void setConsoleOutput(ConsoleOutput consoleOutput) {
         loggingConfiguration.setConsoleOutput(consoleOutput);
     }
 
+    @Override
+    public Set<WarningType> getWarningTypes() {
```
Comment (bmuschko) [inline subprojects/core-api/src/main/java/org/gradle/StartParameter.java]: :x: This should also be reflected in the methods `prepareNewInstance` and `toString`.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy [hunk 3]
```diff
@@ -41,30 +42,33 @@ class CommandLineArgDeprecationIntegrationTest extends AbstractIntegrationSpec {
         'https://github.com/gradle/gradle/issues/1425' | '--recompile-scripts' | RECOMPILE_SCRIPTS_MESSAGE
         'https://github.com/gradle/gradle/issues/3077' | '--no-rebuild'        | NO_REBUILD_MESSAGE
         'https://github.com/gradle/gradle/issues/3077' | '-a'                  | NO_REBUILD_MESSAGE
-        'https://github.com/gradle/gradle/issues/3334' | '--no-search-upward'  | NO_SEARCH_UPWARD_MESSAGE
```
Comment (bmuschko) [inline subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: Why was this removed?

File: subprojects/docs/src/docs/release/notes.md [hunk 4]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
   NOTE: Totally fine to just link to an example that show the feature.
 -->
 
+### Deprecation warning improvements
+
+In this release, deprecation warnings are no longer displayed in the build by default. Instead, all deprecation warnings in the build will be collected and a single "summary" will be rendered at the end of the build:
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: ...no longer displayed **in the console output** by default

File: subprojects/docs/src/docs/release/notes.md [hunk 5]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
...
+
+    There're X deprecation warnings, which may break the build in Gradle 5.0. Please run with --warnings=all to see them.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :o: Instead of `X` I'd use \<number\> to indicate that it is a counter.

File: subprojects/docs/src/docs/release/notes.md [hunk 6]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
...
+
+As said by this warning, you can run the build with command line option `--warnings=all` or property `org.gradle.warnings=all` to have all warnings displayed as before.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :o: I'd completely remove "As said by this warning,"

Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: Wording: "...or the property..."

Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :o: I am not quite sure what we mean by "as before". I guess you mean in earlier versions of Gradle?

File: subprojects/docs/src/docs/release/notes.md [hunk 7]
```diff
@@ -24,6 +24,15 @@ Think of every feature section as a mini blog post.
...
+Of course, you can use command line option `--warnings=none` or property `org.gradle.warnings=none` to suppress all warnings, including the one displayed at the end of the build.
```
Comment (bmuschko) [inline subprojects/docs/src/docs/release/notes.md]: :x: "...the command line option..."
:x: "...the property..."

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 8]
```diff
@@ -735,6 +740,18 @@ public GradleExecuter withOwnUserHomeServices() {
         return this;
     }
 
+    @Override
+    public GradleExecuter withWarnings(WarningType... warningTypes) {
+        this.warningTypes = Sets.newHashSet(warningTypes);
```
Comment (bmuschko) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: :o: Maybe just use `return withWarnings(Sets.newHashSet(warningTypes));`.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 9]
```diff
@@ -860,6 +877,10 @@ private boolean noDaemonArgumentGiven() {
             allArgs.add("--console=" + consoleType.toString().toLowerCase());
         }
 
+        for (WarningType type : warningTypes) {
+            allArgs.add("--warnings=" + type.getBuildOption());
```
Comment (bmuschko) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: :x: What exactly happens if you use multiple warning types? What's the use case? Shouldn't there be just one?

File: subprojects/core-api/src/main/java/org/gradle/StartParameter.java (hunk 2 above)
Comment (bmuschko) [inline subprojects/core-api/src/main/java/org/gradle/StartParameter.java]: :x: Would need to have `@since` and `@Incubating`.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ */
+@Incubating
+public enum WarningType {
+    /**
+     * Show all warnings (e.g. deprecation warnings)
+     */
+    All("all"),
+
+    /**
+     * Default value. By default, all deprecation warnings will be suppressed and a single summary message will be displayed at the end of a build indicating how many deprecation warnings suppressed.
```
Comment (bmuschko) [inline subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java]: :o: The Javadocs here could be clearer. I'd remove "By default, ..." and start with the description before saying that it is the default value.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+import org.gradle.api.Incubating;
+
+/**
+ * Specifies the warning type a user wants to see
+ *
+ * @since 4.5
+ */
+@Incubating
+public enum WarningType {
```
Comment (bmuschko) [inline subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java]: :x: Should be covered by a unit test.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingConfigurationBuildOptions.java [hunk 12]
```diff
@@ -170,4 +177,31 @@ public void applyTo(String value, LoggingConfiguration settings, Origin origin)
...
+    public static class WarningsOption extends ListBuildOption<LoggingConfiguration> {
+        public static final String LONG_OPTION = "warnings";
+        public static final String GRADLE_PROPERTY = "org.gradle.warnings";
+
+        public WarningsOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create(LONG_OPTION, "Specifies which type of warnings to generate. Values are 'all', 'summary' (default) or 'no'"));
+        }
+
+        @Override
+        public void applyTo(List<String> values, LoggingConfiguration settings, final Origin origin) {
```
Comment (bmuschko) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingConfigurationBuildOptions.java]: :x: I'd think that we'll want to have some unit testing for this logic.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java (hunk 1 above)
Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: Thanks @wolfs . Personally I don't like `--show-warnings` and `--warnings` because their names are so similar that users might be puzzled. I think we could have `--warnings` for `all`/`none`/`<category>` for now:

- By default, this set is empty, all categories of warnings are suppressed and a summary will be rendered.
- If special value `none`/`all` is specified, all other values are overwritten, just like @oehme said. This means it's allowed to have `org.gradle.warnings=all` and cmd args `--warnings=none`
- `--warnings=all --warnings=none` is undefined.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy (hunk 3 above)
Comment (blindpirate) [reply subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: That's removed by you... https://github.com/gradle/gradle/pull/3651

File: subprojects/core-api/src/main/java/org/gradle/StartParameter.java (hunk 2 above)
Comment (blindpirate) [reply subprojects/core-api/src/main/java/org/gradle/StartParameter.java]: Added `@inheritDoc`

Comment (blindpirate) [reply subprojects/core-api/src/main/java/org/gradle/StartParameter.java]: Nice catch. Done.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java (hunk 9 above)
Comment (blindpirate) [reply subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: In the long term we may allow `--warnings=deprecation,compiler,<category>` etc. Please see discussion above.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingConfigurationBuildOptions.java (hunk 12 above)
Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/internal/logging/LoggingConfigurationBuildOptions.java]: Done.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java (hunk 1 above)
Comment (wolfs) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I am not very fixed on what the options are called. I would very like to see both concepts modelled differently, though.
IIUC, we currently only implement the way how to show the warnings.
So it is OK to only have the switch for the how to present the warnings, with possible values `summary`, `full` or `none`. That would not be a `Set` but a single value.
Then we can later add support for categories, like `deprecation`, etc.
WDYT?

Comment (bmuschko) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: Before reading this thread I also wondering why we can provide a Set of warnings. Now this makes more sense. I agree with @wolfs that we should separate warning type from the warning category/level.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy (hunk 3 above)
Comment (bmuschko) [reply subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: Ah, kind of weird that it shows up as change in your PR though.

File: subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java [hunk 13]
```diff
@@ -877,9 +873,7 @@ private boolean noDaemonArgumentGiven() {
             allArgs.add("--console=" + consoleType.toString().toLowerCase());
         }
 
-        for (WarningType type : warningTypes) {
-            allArgs.add("--warnings=" + type.toString().toLowerCase(Locale.ENGLISH));
-        }
+        allArgs.add("--warnings=" + warningType.toString().toLowerCase(Locale.ENGLISH));
```
Comment (bmuschko) [inline subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/AbstractGradleExecuter.java]: Why's the `Locale` important here?

**Comment by bmuschko** [comment]
@blindpirate Is this a crucial change for 4.5? If yes, then we should try to merge it before branching and mark the PR with the proper milestone.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java (hunk 11 above)
Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/WarningType.java]: Thanks, now I think UT is no longer needed.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java (hunk 1 above)
Comment (blindpirate) [reply subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: Thanks @wolfs @bmuschko @oehme I've already done this. Please take a look at it again when you're free.

File: subprojects/docs/src/docs/release/notes.md (hunk 4 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Thanks, done.

File: subprojects/docs/src/docs/release/notes.md (hunk 5 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

File: subprojects/docs/src/docs/release/notes.md (hunk 6 above)
Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

Comment (blindpirate) [reply subprojects/docs/src/docs/release/notes.md]: Done.

**Comment by blindpirate** [comment]
@bmuschko Yes, that's for 4.5. I added the milestone but this still needs to be reviewed. I'll merge this into `release` later.

File: subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java [hunk 14]
```diff
@@ -44,6 +45,20 @@
...
 
+    /**
+     * Specifies which type of warnings should be written to the console.
+     * @since 4.5
+     */
+    @Incubating
+    WarningType getWarningType();
+
+    /**
+     * Specifies which type of warnings should be written to the console.
```
Comment (oehme) [inline subprojects/logging/src/main/java/org/gradle/api/logging/configuration/LoggingConfiguration.java]: I don't feel like "warningType" is the right word for this. It's not about types of warnings, but the way they are displayed. Maybe "WarningMode"?

**Comment by blindpirate** [comment]
@oehme Now it's renamed to `WarningMode` and `--warning-mode` because I think `--warnings` is kind of misleading users.
